File: js/moq/src/ietf/subscribe_announces.ts
import type * as Path from "../path.ts";
import type { Reader, Writer } from "../stream.ts";
import * as Namespace from "./namespace.ts";

export class SubscribeAnnounces {
	static id = 0x11;

	namespace: Path.Valid;

	constructor(namespace: Path.Valid) {
		this.namespace = namespace;
	}

	async encodeMessage(w: Writer): Promise<void> {
		await Namespace.encode(w, this.namespace);
		await w.u8(0); // no parameters
	}

	static async decodeMessage(r: Reader): Promise<SubscribeAnnounces> {
		const namespace = await Namespace.decode(r);

		const numParams = await r.u8();
		if (numParams !== 0) {
			throw new Error(`SUBSCRIBE_ANNOUNCES: parameters not supported: ${numParams}`);
		}

		return new SubscribeAnnounces(namespace);
	}
}

export class SubscribeAnnouncesOk {
	static id = 0x12;

	namespace: Path.Valid;

	constructor(namespace: Path.Valid) {
		this.namespace = namespace;
	}

	async encodeMessage(w: Writer): Promise<void> {
		await Namespace.encode(w, this.namespace);
	}

	static async decodeMessage(r: Reader): Promise<SubscribeAnnouncesOk> {
		const namespace = await Namespace.decode(r);
		return new SubscribeAnnouncesOk(namespace);
	}
}

export class SubscribeAnnouncesError {
	static id = 0x13;

	namespace: Path.Valid;
	errorCode: number;
	reasonPhrase: string;

	constructor(namespace: Path.Valid, errorCode: number, reasonPhrase: string) {
		this.namespace = namespace;
		this.errorCode = errorCode;
		this.reasonPhrase = reasonPhrase;
	}

	async encodeMessage(w: Writer): Promise<void> {
		await Namespace.encode(w, this.namespace);
		await w.u62(BigInt(this.errorCode));
		await w.string(this.reasonPhrase);
	}

	static async decodeMessage(r: Reader): Promise<SubscribeAnnouncesError> {
		const namespace = await Namespace.decode(r);
		const errorCode = Number(await r.u62());
		const reasonPhrase = await r.string();

		return new SubscribeAnnouncesError(namespace, errorCode, reasonPhrase);
	}
}

export class UnsubscribeAnnounces {
	static id = 0x14;

	namespace: Path.Valid;

	constructor(namespace: Path.Valid) {
		this.namespace = namespace;
	}

	async encodeMessage(w: Writer): Promise<void> {
		await Namespace.encode(w, this.namespace);
	}

	static async decodeMessage(r: Reader): Promise<UnsubscribeAnnounces> {
		const namespace = await Namespace.decode(r);
		return new UnsubscribeAnnounces(namespace);
	}
}

File: rs/moq/src/ietf/subscribe_announces.rs
//! IETF moq-transport-07 subscribe announces messages

use std::borrow::Cow;

use crate::{coding::*, Path};

use super::util::{decode_namespace, encode_namespace};

/// SubscribeAnnounces message (0x11)
#[derive(Clone, Debug)]
pub struct SubscribeAnnounces<'a> {
	pub namespace: Path<'a>,
}

impl<'a> Message for SubscribeAnnounces<'a> {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		encode_namespace(w, &self.namespace);
		0u8.encode(w); // no parameters
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let namespace = decode_namespace(r)?;

		let num_params = u8::decode(r)?;
		if num_params != 0 {
			return Err(DecodeError::InvalidValue);
		}

		Ok(Self { namespace })
	}
}

/// SubscribeAnnouncesOk message (0x12)
#[derive(Clone, Debug)]
pub struct SubscribeAnnouncesOk<'a> {
	pub namespace: Path<'a>,
}

impl<'a> Message for SubscribeAnnouncesOk<'a> {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		encode_namespace(w, &self.namespace);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let namespace = decode_namespace(r)?;
		Ok(Self { namespace })
	}
}

/// SubscribeAnnouncesError message (0x13)
#[derive(Clone, Debug)]
pub struct SubscribeAnnouncesError<'a> {
	pub namespace: Path<'a>,
	pub error_code: u64,
	pub reason_phrase: Cow<'a, str>,
}

impl<'a> Message for SubscribeAnnouncesError<'a> {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		encode_namespace(w, &self.namespace);
		self.error_code.encode(w);
		self.reason_phrase.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let namespace = decode_namespace(r)?;
		let error_code = u64::decode(r)?;
		let reason_phrase = Cow::<str>::decode(r)?;

		Ok(Self {
			namespace,
			error_code,
			reason_phrase,
		})
	}
}

/// UnsubscribeAnnounces message (0x14)
#[derive(Clone, Debug)]
pub struct UnsubscribeAnnounces<'a> {
	pub namespace: Path<'a>,
}

impl<'a> Message for UnsubscribeAnnounces<'a> {
	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		encode_namespace(w, &self.namespace);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let namespace = decode_namespace(r)?;
		Ok(Self { namespace })
	}
}
